pkg/planner/core/planbuilder.go

```diff
@@ -5419,6 +5446,97 @@ func (b *PlanBuilder) buildDDL(ctx context.Context, node ast.DDLNode) (base.Plan
 	return p, nil
 }
 
+// buildColsFromPlan builds the columns of the table from the logical plan
+// used in create table as select statement
+func buildColsFromPlan(v *ast.CreateTableStmt, logicalPlan base.LogicalPlan) error {
+	schema := logicalPlan.Schema()
+	names := logicalPlan.OutputNames()
+
+	// fill select columns,used in build sql when insert into table(v.SelectColumns...)  select ...
+	for _, name := range names {
+		v.SelectColumns = append(v.SelectColumns, &name.ColName)
+	}
+	if v.Cols == nil {
+		// Order of columns is the same as the order of [cols from select stmt]
+		v.Cols = make([]*ast.ColumnDef, len(schema.Columns))
+		for i, name := range names {
+			checkRetType(schema.Columns[i].RetType)
+			v.Cols[i] = &ast.ColumnDef{
+				Name: &ast.ColumnName{
+					Name: name.ColName,
+				},
+				Tp:      schema.Columns[i].RetType,
+				Options: []*ast.ColumnOption{},
+			}
+		}
+	} else {
+		// Compatible with MySQL order of columns
+		// colName -> (namesIndex,ColsIndex)
+		// if IndexArr[1] != -1,it means the column is explicitly specified when creating the table, then the column is used.
+		colMap := make(map[string][2]int)
+		for i, name := range names {
+			colMap[name.ColName.L] = [2]int{i, -1}
+		}
+		// First: check col which is explicitly specified and not in the [cols from select stmt],add it to the newCols
+		newCols := make([]*ast.ColumnDef, 0, len(schema.Columns))
+		for i, col := range v.Cols {
+			if arr, exists := colMap[col.Name.Name.L]; !exists {
+				// if NOT NULL is set,throw error
+				for _, opt := range col.Options {
+					if opt.Tp == ast.ColumnOptionNotNull {
+						return table.ErrNoDefaultValue.GenWithStackByArgs(col.Name.Name.L)
+					}
+				}
+				newCols = append(newCols, col)
+			} else {
+				// The type which is explicitly specified is used as the basis
+				arr[1] = i
+				colMap[col.Name.Name.L] = arr
+			}
+		}
+		// Second: add the remaining cols to the newCols
+		for i, name := range names {
+			// If the column is explicitly specified when creating the table
+			if colMap[name.ColName.L][1] != -1 {
+				newCols = append(newCols, v.Cols[colMap[name.ColName.L][1]])
+			} else {
+				checkRetType(schema.Columns[i].RetType)
+				newCols = append(newCols, &ast.ColumnDef{
+					Name: &ast.ColumnName{
+						Name: name.ColName,
+					},
+					Tp: schema.Columns[i].RetType,
+				})
+			}
+		}
+		v.Cols = newCols
+	}
+	return nil
+}
+
+// checkRetType checks the type of the column and changes it (eg. VarString -> Varchar)
+func checkRetType(fieldType *types.FieldType) {
```

Comment:
the function name check... seems will not modify the arguments. Maybe adjust/convert or something